cmd/zlint-mozilla-trusted-roots-update/main.go

```diff
@@ -0,0 +1,211 @@
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"go/format"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"text/template"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/zmap/zcrypto/x509"
+)
+
+const MOZILLA_CERTDATA = "http://mxr.mozilla.org/mozilla/source/security/nss/lib/ckfw/builtins/certdata.txt?raw=1"
+
+var (
+	// httpClient is a http.Client instance configured with timeouts.
+	httpClient = &http.Client{
+		Transport: &http.Transport{
+			Dial: (&net.Dialer{
+				Timeout:   15 * time.Second,
+				KeepAlive: 15 * time.Second,
+			}).Dial,
+			TLSHandshakeTimeout:   5 * time.Second,
+			ResponseHeaderTimeout: 5 * time.Second,
+			ExpectContinueTimeout: 1 * time.Second,
+		},
+	}
+	// mozillaTrustedRootsTemplate is a template that produces a Golang source
+	// code file in the "util" package containing a single member variable,
+	// a slice of strings containing the hex encoded SPKIs of all roots
+	// trusted by Mozilla.
+	mozillaTrustedRootsTemplate = template.Must(template.New(
+		"mozillaTrustedRootsTemplate").Parse(
+		`// Code generated by go generate; DO NOT EDIT.
+// This file was generated by zlint-mozilla-trusted-roots-update.
+
+/*
+ * ZLint Copyright 2018 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package util
+
+var mozillaTrustedSPKIs = []string{
+{{- range . }}
+	"{{ . }}",
+{{- end }}
+}
+`))
+)
+
+func getMozillaTrustedCerts() ([]*x509.Certificate, error) {
+	resp, err := httpClient.Get(MOZILLA_CERTDATA)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch data from %q : %s",
+			MOZILLA_CERTDATA, err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status code fetching data "+
+			"from %q : expected status %d got %d",
+			MOZILLA_CERTDATA, http.StatusOK, resp.StatusCode)
+	}
+
+	var trustedCertificates []*x509.Certificate
```

Comment:
The scanning logic repeats some steps, nests quite deep and doesn't check errors

Alternative could be:

```
var trustedCertificates []*x509.Certificate
	scanner := bufio.NewScanner(resp.Body)
	var readLine bool
	var certdata []byte
	for scanner.Scan() {
		line := scanner.Text()
		line = strings.ReplaceAll(line, "\r\n", "")
		
		if line == "CKA_VALUE MULTILINE_OCTAL" {
			readLine = true
			continue
		}else if line == "END"{
			readLine = false
			continue
		}else if !readLine {
			continue
		}
		
		bs := strings.Split(line, "\\")
		for _, b := range bs {
			if b == "" {
				continue
			}
			i, err := strconv.ParseInt(b, 8, 0)
			if err != nil {
				return nil, fmt.Errorf(
					"unexpected octal character : %s", err)
			}
			certdata = append(certdata, byte(i))
		}	
	}

        if err:= scanner.Err(); err != nil {
		return nil, fmt.Errorf("error reading %q: %s",MOZILLA_CERTDATA,err)
	}

	cert, err := x509.ParseCertificate(certdata)
	if err != nil {
		return nil, fmt.Errorf("unable to parse certificate : %s", err)
	}

	trustedCertificates = append(trustedCertificates, cert)
```

Reply:
Well, it's just an extra level of nesting which repeats the reading of a line. Not sure that it's worth the change since IMHO the current structure is easier to read.

Reply:
Echoing my [earlier comment](https://github.com/zmap/zlint/pull/323#discussion_r337074333) I would still prefer to see this utility and the review of it be separated into a second PR if possible.

Reply:
I have already removed it from this PR. As soon as it gets accepted, I will open a new one.

Reply:
D'oh! Sorry @fotisl, I missed that last commit. Thank you!